File: profiling/workload.py
#!/usr/bin/env python3
"""
Generates test workload against a running Chromia node.
Sends transactions via `chr` CLI and queries via REST API.

Usage: ./workload.py [--chr PATH] [--users N] [--posts N] [--node-url URL] [--results PATH]
"""

import argparse
import concurrent.futures
import json
import os
import subprocess
import sys
import time
from pathlib import Path
from urllib.error import URLError
from urllib.request import Request, urlopen

SCRIPT_DIR = Path(__file__).resolve().parent
DAPP_DIR = SCRIPT_DIR / "dapp"
DEFAULT_CHR = SCRIPT_DIR.parent / "chromia-cli-local" / "chromia-cli" / "target" / "chromia-cli-dev-dist" / "bin" / "chr"


def log(msg: str) -> None:
    ts = time.strftime("%H:%M:%S", time.gmtime())
    print(f"[workload] {ts} {msg}", flush=True)


def wait_for_node(url: str, timeout: int = 60) -> None:
    log(f"Waiting for node at {url} ...")
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            urlopen(f"{url}/brid/iid_0", timeout=2)
            log("Node is ready")
            return
        except (URLError, OSError):
            time.sleep(1)
    sys.exit(f"Node did not become ready after {timeout}s")


def get_brid(url: str) -> str:
    resp = urlopen(f"{url}/brid/iid_0", timeout=5)
    return resp.read().decode().strip().strip('"')


def chr_tx(chr_bin: str, node_url: str, brid: str, op: str, *args: str) -> int:
    """Submit one transaction via chr. Returns return-code.
    --await blocks until the tx is included in a block.
    Note: chr accepts -brid (single dash), not --brid."""
    # chr CLI expects each Rell argument as a GTV literal: text values
    # must be quoted, numbers are bare. Assume text args for now.
    gtv_args = [f'"{a}"' for a in args]
    cmd = [chr_bin, "tx", "--await", "--api-url", node_url, "-brid", brid,
           op, *gtv_args]
    r = subprocess.run(cmd, cwd=str(DAPP_DIR), capture_output=True, timeout=60)
    return r.returncode


def chr_tx_parallel(chr_bin: str, node_url: str, brid: str,
                    tasks: list[tuple[str, tuple[str, ...]]],
                    workers: int) -> int:
    """Submit many transactions concurrently. With many parallel
    submissions, multiple transactions land in the same block — the
    node stops sitting idle between blocks.

    Returns the number of failures.
    """
    failures = 0
    with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as pool:
        futures = [pool.submit(chr_tx, chr_bin, node_url, brid, op, *args)
                   for op, args in tasks]
        for f in concurrent.futures.as_completed(futures):
            if f.result() != 0:
                failures += 1
    return failures


def rest_query(node_url: str, body: dict) -> bytes:
    data = json.dumps(body).encode()
    req = Request(f"{node_url}/query/iid_0", data=data, headers={"Content-Type": "application/json"})
    try:
        return urlopen(req, timeout=10).read()
    except (URLError, OSError):
        return b'{"error":"query failed"}'


def timed(fn) -> float:
    start = time.monotonic()
    fn()
    return round(time.monotonic() - start, 3)


def main() -> None:
    parser = argparse.ArgumentParser(description="Rell profiling workload generator")
    parser.add_argument("--chr", default=str(DEFAULT_CHR), help="Path to chr binary")
    parser.add_argument("--users", type=int, default=20, help="Number of test users")
    parser.add_argument("--posts", type=int, default=10, help="Posts per user")
    parser.add_argument("--node-url", default=os.environ.get("NODE_URL", "http://localhost:7740"))
    parser.add_argument("--results", default=None, help="Path for results JSON")
    parser.add_argument("--tx-workers", type=int, default=16,
                        help="Concurrent chr tx submitters (default: 16). "
                             "Higher values pack more tx per block and keep "
                             "the node from idling between blocks.")
    args = parser.parse_args()

    results_path = Path(args.results) if args.results else SCRIPT_DIR / "reports" / "workload-results.json"
    results_path.parent.mkdir(parents=True, exist_ok=True)

    wait_for_node(args.node_url)
    brid = get_brid(args.node_url)
    log(f"Blockchain RID: {brid}")

    total_start = time.monotonic()

    workers = args.tx_workers
    log(f"Using {workers} parallel tx submitters")

    # Phase 1: Create users (parallel — fills blocks instead of one-per-block)
    log(f"Phase 1: Creating {args.users} users...")
    tasks = [("create_user",
              (f"user_{i}", f"Bio for user {i} — blockchain developer and tester"))
             for i in range(1, args.users + 1)]
    def phase1():
        fails = chr_tx_parallel(args.chr, args.node_url, brid, tasks, workers)
        if fails: log(f"  WARNING: {fails}/{len(tasks)} tx failed")
    p1_time = timed(phase1)
    log(f"  Created {args.users} users in {p1_time}s")

    # Phase 2: Create posts (parallel)
    total_posts = args.users * args.posts
    log(f"Phase 2: Creating {total_posts} posts ({args.posts} per user)...")
    tasks = [("create_post",
              (f"user_{i}", f"Post {j} by user {i}",
               f"Content of post {j} by user_{i}. "
               f"Exercises database storage and retrieval."))
             for i in range(1, args.users + 1)
             for j in range(1, args.posts + 1)]
    def phase2():
        fails = chr_tx_parallel(args.chr, args.node_url, brid, tasks, workers)
        if fails: log(f"  WARNING: {fails}/{len(tasks)} tx failed")
    p2_time = timed(phase2)
    log(f"  Created {total_posts} posts in {p2_time}s")

    # Phase 3: Update bios (parallel)
    log(f"Phase 3: Updating {args.users} user bios...")
    tasks = [("update_bio", (f"user_{i}", f"Updated bio #{i} — now with more detail"))
             for i in range(1, args.users + 1)]
    def phase3():
        fails = chr_tx_parallel(args.chr, args.node_url, brid, tasks, workers)
        if fails: log(f"  WARNING: {fails}/{len(tasks)} tx failed")
    p3_time = timed(phase3)
    log(f"  Updated {args.users} bios in {p3_time}s")

    # Phase 4: Queries via REST API (parallel — saturates the query path)
    log("Phase 4: Running queries...")
    QUERY_ROUNDS = 10

    def build_queries():
        qs = []
        for _ in range(QUERY_ROUNDS):
            qs.append({"type": "count_users"})
            qs.append({"type": "count_posts"})
            for i in range(1, args.users + 1):
                qs.append({"type": "get_user", "name": f"user_{i}"})
            for page in range(5):
                qs.append({"type": "list_users", "page_size": 5, "skip": page * 5})
            for i in range(1, args.users + 1):
                qs.append({"type": "get_posts_by_user",
                           "author_name": f"user_{i}", "max_count": args.posts})
            for term in ["Post 1", "Post 5", "content", "blockchain", "user_1"]:
                qs.append({"type": "search_posts", "term": term})
            for i in range(1, args.users + 1):
                qs.append({"type": "get_user_post_count", "author_name": f"user_{i}"})
        return qs

    queries = build_queries()
    query_count = len(queries)

    def phase4():
        with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as pool:
            list(pool.map(lambda q: rest_query(args.node_url, q), queries))

    p4_time = timed(phase4)
    log(f"  Ran {query_count} queries in {p4_time}s")

    total_time = round(time.monotonic() - total_start, 3)
    log(f"Workload complete in {total_time}s")

    results = {
        "total_time_s": total_time,
        "num_users": args.users,
        "posts_per_user": args.posts,
        "total_posts": total_posts,
        "total_queries": query_count,
        "phases": {
            "create_users": {"count": args.users, "time_s": p1_time},
            "create_posts": {"count": total_posts, "time_s": p2_time},
            "update_bios":  {"count": args.users, "time_s": p3_time},
            "queries":      {"count": query_count, "time_s": p4_time},
        },
    }
    results_path.write_text(json.dumps(results, indent=2))
    log(f"Results written to {results_path}")


if __name__ == "__main__":
    main()

File: performance/src/main/kotlin/profiler/workload.kt
/*
 * Copyright (C) 2026 ChromaWay AB. See LICENSE for license information.
 */
@file:JvmName("WorkloadKt")

package net.postchain.rell.performance.profiler

import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.databind.node.JsonNodeFactory
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.Context
import com.github.ajalt.clikt.core.main
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.types.int
import com.github.ajalt.clikt.parameters.types.path
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.io.path.createDirectories
import kotlin.io.path.div
import kotlin.io.path.inputStream
import kotlin.io.path.outputStream
import kotlin.system.exitProcess
import kotlin.system.measureNanoTime

class WorkloadCommand: CliktCommand(name = "workload") {
    override fun help(context: Context) =
        "Generate test transactions and queries against a running Chromia node."

    private val chrBin by option("--chr", help = "Path to chr binary").default(defaultChrBin())
    private val users by option("--users", help = "Number of test users").int().default(20)
    private val postsPerUser by option("--posts", help = "Posts per user").int().default(10)
    private val nodeUrl by option("--node-url", help = "Postchain REST API URL")
        .default(System.getenv("NODE_URL") ?: "http://localhost:7740")
    private val resultsPath by option("--results", help = "Output path for results JSON")
        .path().default(perfDir() / "reports" / "workload-results.json")
    private val txWorkers by option(
        "--tx-workers",
        help = "Concurrent chr tx submitters (higher values pack more tx per block)",
    )
        .int().default(16)

    override fun run() {
        resultsPath.parent.createDirectories()
        waitForNode(nodeUrl)
        val brid = fetchBrid(nodeUrl)
        log("workload", "Blockchain RID: $brid")
        log("workload", "Using $txWorkers parallel tx submitters")

        val totalNanos = measureNanoTime {
            val (p1Tasks, p1Time) = run {
                val tasks = (1..users).map { i ->
                    "create_user" to listOf("user_$i", "Bio for user $i — blockchain developer and tester")
                }
                log("workload", "Phase 1: Creating $users users...")
                val t = timed { runTxBatch(tasks) }
                log("workload", "  Created $users users in ${"%.3f".format(t)}s")
                tasks.size to t
            }

            val totalPosts = users * postsPerUser
            val (p2Tasks, p2Time) = run {
                val tasks = (1..users).flatMap { i ->
                    (1..postsPerUser).map { j ->
                        "create_post" to listOf(
                            "user_$i",
                            "Post $j by user $i",
                            "Content of post $j by user_$i. Exercises database storage and retrieval.",
                        )
                    }
                }
                log("workload", "Phase 2: Creating $totalPosts posts ($postsPerUser per user)...")
                val t = timed { runTxBatch(tasks) }
                log("workload", "  Created $totalPosts posts in ${"%.3f".format(t)}s")
                tasks.size to t
            }

            val (p3Tasks, p3Time) = run {
                val tasks = (1..users).map { i ->
                    "update_bio" to listOf("user_$i", "Updated bio #$i — now with more detail")
                }
                log("workload", "Phase 3: Updating $users user bios...")
                val t = timed { runTxBatch(tasks) }
                log("workload", "  Updated $users bios in ${"%.3f".format(t)}s")
                tasks.size to t
            }

            val queries = buildQueries(users, postsPerUser, rounds = 10)
            log("workload", "Phase 4: Running queries...")
            val p4Time = timed { runQueriesBatch(queries) }
            log("workload", "  Ran ${queries.size} queries in ${"%.3f".format(p4Time)}s")

            writeResults(
                totalTimeS = -1.0,
                phases = mapOf(
                    "create_users" to (p1Tasks to p1Time),
                    "create_posts" to (p2Tasks to p2Time),
                    "update_bios" to (p3Tasks to p3Time),
                    "queries" to (queries.size to p4Time),
                ),
                totalPosts = totalPosts,
            )
        }
        val totalSeconds = totalNanos / 1_000_000_000.0

        rewriteWithTotal(totalSeconds)
        log("workload", "Workload complete in ${"%.3f".format(totalSeconds)}s")
        log("workload", "Results written to $resultsPath")
    }

    private fun runTxBatch(tasks: List<Pair<String, List<String>>>) {
        val pool = Executors.newFixedThreadPool(txWorkers)
        try {
            val futures = tasks.map { (op, args) -> pool.submit(Callable { submitTx(op, args) }) }
            var failures = 0
            for (f in futures) if (!f.get()) failures++
            if (failures > 0) log("workload", "  WARNING: $failures/${tasks.size} tx failed")
        } finally {
            pool.shutdown()
            pool.awaitTermination(5, TimeUnit.MINUTES)
        }
    }

    private fun submitTx(op: String, args: List<String>): Boolean {
        // chr CLI expects each Rell argument as a GTV literal — text values must be quoted.
        val gtvArgs = args.map { "\"$it\"" }
        val brid = bridCache
            ?: error("BRID must be fetched before submitting transactions")
        val cmd = mutableListOf(chrBin, "tx", "--await", "--api-url", nodeUrl, "-brid", brid, op)
        cmd.addAll(gtvArgs)
        return try {
            val proc = ProcessBuilder(cmd)
                .directory((perfDir() / "dapp").toFile())
                .redirectErrorStream(true).start()
            val finished = proc.waitFor(60, TimeUnit.SECONDS)
            if (!finished) {
                proc.destroyForcibly()
                false
            } else {
                proc.exitValue() == 0
            }
        } catch (_: Exception) {
            false
        }
    }

    private val mapper = jacksonObjectMapper().configure(SerializationFeature.INDENT_OUTPUT, true)
    private val httpClient: HttpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(10))
        .build()

    private var bridCache: String? = null

    private fun fetchBrid(url: String): String {
        val req = HttpRequest.newBuilder(URI("$url/brid/iid_0")).GET()
            .timeout(Duration.ofSeconds(5)).build()
        val resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
        require(resp.statusCode() == 200) { "Failed to fetch BRID: HTTP ${resp.statusCode()}" }
        val brid = resp.body().trim().trim('"')
        bridCache = brid
        return brid
    }

    private fun waitForNode(url: String, timeoutSec: Long = 60) {
        log("workload", "Waiting for node at $url ...")
        val deadline = System.nanoTime() + timeoutSec * 1_000_000_000L
        while (System.nanoTime() < deadline) {
            try {
                val req = HttpRequest.newBuilder(URI("$url/brid/iid_0"))
                    .timeout(Duration.ofSeconds(2)).GET().build()
                if (httpClient.send(req, HttpResponse.BodyHandlers.discarding()).statusCode() == 200) {
                    log("workload", "Node is ready")
                    return
                }
            } catch (_: Exception) {
                Thread.sleep(1000)
            }
        }
        exitProcess(1)
    }

    private fun runQueriesBatch(queries: List<ObjectNode>) {
        val pool: ExecutorService = Executors.newFixedThreadPool(txWorkers)
        try {
            val futures = queries.map { q ->
                pool.submit { runQuery(q) }
            }

            for (f in futures) f.get()
        } finally {
            pool.shutdown()
            pool.awaitTermination(5, TimeUnit.MINUTES)
        }
    }

    private fun runQuery(body: ObjectNode) {
        try {
            val req = HttpRequest.newBuilder(URI("$nodeUrl/query/iid_0"))
                .timeout(Duration.ofSeconds(10))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body)))
                .build()

            httpClient.send(req, HttpResponse.BodyHandlers.discarding())
        } catch (_: Exception) {}
    }

    @Suppress("SameParameterValue")
    private fun buildQueries(users: Int, postsPerUser: Int, rounds: Int): List<ObjectNode> {
        val nf: JsonNodeFactory = JsonNodeFactory.instance
        val out = ArrayList<ObjectNode>(rounds * (2 + users + 5 + users + 5 + users))

        repeat(rounds) {
            out += nf.objectNode().put("type", "count_users")
            out += nf.objectNode().put("type", "count_posts")

            for (i in 1..users)
                out += nf.objectNode().put("type", "get_user").put("name", "user_$i")

            for (page in 0 until 5)
                out += nf.objectNode().put("type", "list_users").put("page_size", 5).put("skip", page * 5)

            for (i in 1..users)
                out += nf.objectNode()
                    .put("type", "get_posts_by_user")
                    .put("author_name", "user_$i")
                    .put("max_count", postsPerUser)

            for (term in listOf("Post 1", "Post 5", "content", "blockchain", "user_1"))
                out += nf.objectNode()
                    .put("type", "search_posts")
                    .put("term", term)

            for (i in 1..users)
                out += nf.objectNode()
                    .put("type", "get_user_post_count")
                    .put("author_name", "user_$i")
        }
        return out
    }

    @Suppress("SameParameterValue")
    private fun writeResults(
        totalTimeS: Double,
        phases: Map<String, Pair<Int, Double>>,
        totalPosts: Int,
    ) {
        val nf: JsonNodeFactory = JsonNodeFactory.instance
        val root: ObjectNode = nf.objectNode().apply {
            put("total_time_s", totalTimeS)
            put("num_users", users)
            put("posts_per_user", postsPerUser)
            put("total_posts", totalPosts)
            val phasesNode = nf.objectNode()
            for ((name, p) in phases) {
                val (count, t) = p
                phasesNode.set<ObjectNode>(name, nf.objectNode().put("count", count).put("time_s", t))
            }
            set<ObjectNode>("phases", phasesNode)
        }
        mapper.writeValue(resultsPath.outputStream(), root)
    }

    private fun rewriteWithTotal(totalSeconds: Double) {
        val tree = mapper.readTree(resultsPath.inputStream()) as ObjectNode
        tree.put("total_time_s", totalSeconds)
        mapper.writeValue(resultsPath.outputStream(), tree)
    }
}

private inline fun timed(block: () -> Unit): Double = measureNanoTime(block) / 1_000_000_000.0

private fun defaultChrBin(): String = (repoRoot() / "chromia-cli-local" / "chromia-cli" / "target" /
        "chromia-cli-dev-dist" / "bin" / "chr").toString()

fun main(args: Array<String>) = WorkloadCommand().main(args)
